### PR TITLE
Unify grappler optimizer and rewrite-pass

### PR DESCRIPTION
**Pull request by adk9**
* Remove duplicated code in ngraph-optimizer and make ngraph-optimizer use
 rewrite-pass.
* Remove unnecessary NgraphEncapsulationPass
* Move ngraph-optimizer.{cc,h} to ngraph_bridge/

### CODE REVIEW

File: ngraph_bridge/ngraph_rewrite_pass.h [hunk 1]
```diff
@@ -13,14 +13,36 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  *******************************************************************************/
+
 #pragma once
 
-#include "tensorflow/core/graph/graph.h"
+#include "tensorflow/core/common_runtime/optimization_registry.h"
+
+#include <iomanip>
```
Comment (kanvi-nervana) [inline ngraph_bridge/ngraph_rewrite_pass.h:21]: Minor: Move std incl. before TF incl.

File: ngraph_bridge/ngraph_optimizer.cc [hunk 2]
```diff
@@ -26,16 +25,14 @@
...
+#include "api.h"
+#include "assign_clusters.h"
+#include "cluster_manager.h"
+#include "deassign_clusters.h"
+#include "encapsulate_clusters.h"
+#include "log.h"
 #include "ngraph_optimizer.h"
+#include "ngraph_rewrite_pass.h"
 
 #include <iostream>
```
Comment (kanvi-nervana) [inline ngraph_bridge/ngraph_optimizer.cc:37]: Minor: std incl before TF incl

File: ngraph_bridge/ngraph_rewrite_pass.h (hunk 1 above)
Comment (adk9) [reply ngraph_bridge/ngraph_rewrite_pass.h]: Done.

File: ngraph_bridge/ngraph_optimizer.cc (hunk 2 above)
Comment (adk9) [reply ngraph_bridge/ngraph_optimizer.cc]: Done.

File: ngraph_bridge/ngraph_rewrite_pass.cc [hunk 3]
```diff
@@ -64,70 +49,71 @@ mutex NGraphRewritePass::s_serial_counter_mutex;
...
+  // If we don't get a main graph, log that fact and bail.
+  if (options.graph == nullptr) {
+    NGRAPH_VLOG(0) << "NGraphRewritePass: options.graph == nullptr";
     return Status::OK();
   }
-};
+
+  // For filename generation purposes, grab a fresh index. This is just an
+  // arbitrary integer to avoid filename collisions resulting from subsequent
+  // runs of this pass.
```
Comment (kanvi-nervana) [inline ngraph_bridge/ngraph_rewrite_pass.cc:109]: This comment does not look right.

File: ngraph_bridge/ngraph_rewrite_pass.h [hunk 4]
```diff
@@ -13,14 +13,34 @@
...
 namespace tensorflow {
 namespace ngraph_bridge {
 
-Status AddIdentityN(Graph* graph, std::set<string> skip_these_nodes);
+class NGraphRewritePass : public GraphOptimizationPass {
+ public:
+  NGraphRewritePass() = default;
+  ~NGraphRewritePass() override = default;
+
+  Status Rewrite(Graph* graph, std::set<string> skip_these_nodes = {},
```
Comment (kanvi-nervana) [inline ngraph_bridge/ngraph_rewrite_pass.h:29]: Is the separate Rewrite method needed to share code with optimizer? Is it not possible to use the Run method?

**Review by kanvi-nervana — CHANGES_REQUESTED** [review]

File: ngraph_bridge/ngraph_rewrite_pass.h (hunk 4 above)
Comment (adk9) [reply ngraph_bridge/ngraph_rewrite_pass.h:29]: Yes, and no. Please take a look at the definition of the two functions.

File: ngraph_bridge/ngraph_rewrite_pass.cc (hunk 3 above)
Comment (adk9) [reply ngraph_bridge/ngraph_rewrite_pass.cc]: Fixed.

**Review by kanvi-nervana — APPROVED** [review]

**Pull request merged**